File: src/services/GeminiService.ts
/**
 * GeminiService - Google Gemini API
 * 
 * Models: gemini-2.5-flash-lite (primary), gemini-2.5-flash (fallback)
 * Retired: gemini-1.5-flash, gemini-pro, gemini-2.0-flash
 */

import type { FoodEntry } from '../models/types';

const getApiKey = (): string | null => {
  const envKey = import.meta.env.VITE_GEMINI_API_KEY;
  if (envKey && envKey.length > 10 && !envKey.includes('YOUR_')) return envKey;
  const localKey = localStorage.getItem('lifelog_gemini_api_key');
  if (localKey && localKey.length > 10) return localKey;
  return null;
};

export const setGeminiApiKey = (key: string): void => { localStorage.setItem('lifelog_gemini_api_key', key); };
export const hasGeminiApiKey = (): boolean => getApiKey() !== null;
export const isApiKeyFromEnv = (): boolean => {
  const envKey = import.meta.env.VITE_GEMINI_API_KEY;
  return Boolean(envKey && envKey.length > 10 && !envKey.includes('YOUR_'));
};

const BASE = 'https://generativelanguage.googleapis.com/v1beta';
const MODELS = ['gemini-2.5-flash-lite', 'gemini-2.5-flash'];

async function callGemini(prompt: string): Promise<string> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error('API key not configured.');
  let lastError = '';
  for (const model of MODELS) {
    try {
      const ctrl = new AbortController();
      const t = setTimeout(() => ctrl.abort(), 25000);
      const res = await fetch(`${BASE}/models/${model}:generateContent?key=${apiKey}`, {
        method: 'POST', headers: { 'Content-Type': 'application/json' }, signal: ctrl.signal,
        body: JSON.stringify({ contents: [{ parts: [{ text: prompt }] }], generationConfig: { temperature: 0.4, maxOutputTokens: 2048 } }),
      });
      clearTimeout(t);
      const data = await res.json();
      if (data.error) { lastError = `${model}: ${data.error.message}`; continue; }
      const text = data.candidates?.[0]?.content?.parts?.[0]?.text;
      if (text) return text;
      lastError = `${model}: empty response`;
    } catch (e) {
      lastError = e instanceof DOMException && e.name === 'AbortError' ? `${model}: timeout` : (e instanceof Error ? e.message : 'Network error');
    }
  }
  throw new Error(lastError || 'All models failed');
}

async function callGeminiVision(prompt: string, imageBase64: string): Promise<string> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error('API key not configured.');
  const clean = imageBase64.replace(/^data:image\/\w+;base64,/, '');
  for (const model of MODELS) {
    try {
      const ctrl = new AbortController();
      const t = setTimeout(() => ctrl.abort(), 30000);
      const res = await fetch(`${BASE}/models/${model}:generateContent?key=${apiKey}`, {
        method: 'POST', headers: { 'Content-Type': 'application/json' }, signal: ctrl.signal,
        body: JSON.stringify({ contents: [{ parts: [{ text: prompt }, { inline_data: { mime_type: 'image/jpeg', data: clean } }] }] }),
      });
      clearTimeout(t);
      const data = await res.json();
      if (data.error) continue;
      const text = data.candidates?.[0]?.content?.parts?.[0]?.text;
      if (text) return text;
    } catch { continue; }
  }
  throw new Error('OCR failed');
}

/** Result from classifyLog — can contain MULTIPLE structured entries */
export interface ClassifyResult {
  entries: ClassifiedEntry[];
  summary: string;
}

export interface ClassifiedEntry {
  type: 'sleep' | 'exercise' | 'expense' | 'food' | 'activity' | 'general';
  // Sleep
  sleepHours?: number;
  sleepQuality?: string;
  bedTime?: string;
  wakeTime?: string;
  // Exercise
  exerciseDetails?: string;
  exercises?: { name: string; sets?: number; reps?: number; weight?: string }[];
  durationMinutes?: number;
  caloriesBurned?: number;
  // Activity/Steps
  steps?: number;
  distanceKm?: number;
  // Expense
  amount?: number;
  expenseCategory?: string;
  // Food
  foodItems?: { name: string; portion: string; calories: number; protein?: number; carbs?: number; fat?: number; fiber?: number }[];
  mealType?: string;
  totalCalories?: number;
  nutritionSummary?: string;
  // Date
  date?: string; // YYYY-MM-DD if mentioned
}

export const GeminiService = {
  /**
   * THE CORE FUNCTION: Parse natural language into structured life data.
   * Can return MULTIPLE entries from a single log.
   * E.g. "walked 5km (7000 steps) then ate rice" → activity entry + food entry
   */
  async classifyLog(text: string): Promise<ClassifyResult> {
    if (!hasGeminiApiKey()) return mockClassify(text);
    try {
      const todayStr = new Date().toISOString().split('T')[0];
      const prompt = `You are a personal life data parser. Parse this log entry into structured data.
IMPORTANT: A single entry can contain MULTIPLE activities. Extract ALL of them.

Entry: "${text}"
Today's date: ${todayStr}

Return ONLY valid JSON (no markdown, no backticks):
{
  "entries": [
    // Each distinct activity gets its own entry. Examples:
    // For sleep: {"type":"sleep","sleepHours":6.5,"sleepQuality":"good","bedTime":"4:30 AM","wakeTime":"11:00 AM"}
    // For steps/walking: {"type":"activity","steps":5022,"distanceKm":3.71,"date":"2026-05-08"}
    // For exercise: {"type":"exercise","exercises":[{"name":"pushups","sets":3,"reps":10},{"name":"crunches","sets":3,"reps":15}],"durationMinutes":45,"caloriesBurned":320,"exerciseDetails":"pushups 3x10, crunches 3x15"}
    // For expense: {"type":"expense","amount":300,"expenseCategory":"Transportation"}
    // For food: {"type":"food","mealType":"breakfast","totalCalories":650,"nutritionSummary":"High protein breakfast with complex carbs","foodItems":[{"name":"Kala chana","portion":"1 bowl","calories":250,"protein":15,"carbs":30,"fat":5},{"name":"Halwa","portion":"1 serving","calories":300,"protein":3,"carbs":45,"fat":12}]}
  ],
  "summary": "Brief one-line summary of what was logged"
}

CRITICAL RULES:

DATE HANDLING:
- If user mentions ANY date like "10th May 2026", "on 8th May", "yesterday", "day before yesterday", set "date" as "YYYY-MM-DD".
- "yesterday" = calculate yesterday from today (${todayStr}).
- "day before yesterday" = today minus 2 days.
- If NO date mentioned, OMIT the date field (defaults to today).
- Examples: "139₹ jio recharge on 10th May 2026" → date: "2026-05-10"

SLEEP RULES:
- Calculate hours PRECISELY from times. "2 AM to 10 AM" = 8h. "3:50 PM to 5 PM" = 1.17h. "11 PM to 7 AM" = 8h.
- EVERY sleep entry is ADDITIVE. The app sums all sleep entries per day. So "slept 2 AM to 10 AM" = 8h entry, and later "nap 3:50 PM to 5 PM" = 1.17h entry. Total becomes 9.17h.
- NEVER try to "replace" a previous sleep entry. Just create a new one — the app adds them.
- Always include bedTime and wakeTime strings for display.
- Set sleepQuality: if hours >= 7 → "good", if >= 8.5 → "excellent", if < 5 → "poor", else "fair".

FOOD RULES:
- ALWAYS estimate calories, protein, carbs, fat, sugar, fiber for every item.
- Use accurate Indian food nutrition data.
- Each new food entry is ADDITIVE — the app sums calories/protein/etc across all meals.
- Set mealType from context: morning → breakfast, afternoon → lunch, evening → dinner, else → snack.
- nutritionSummary should compare against daily targets if known (e.g. "You've consumed 60% of daily protein target").

EXPENSE RULES:
- Each expense is ADDITIVE — the app sums all expenses per day.
- Extract amount in INR. "₹300 cab" → amount: 300.
- Guess category: food/cab/auto/uber → "Transportation", swiggy/zomato → "Food & Dining", jio/airtel/recharge → "Recharge", amazon/flipkart → "Shopping", electricity/gas/water → "Bills & Utilities".

ACTIVITY/STEPS RULES:
- Each entry is ADDITIVE — the app sums steps per day.
- "walked 0.5 km till 11 AM" = 650 steps, 0.5 km. Later "walked 1 km in evening" = 1300 steps, 1 km. App shows total 1950 steps.
- ALWAYS estimate steps from km: 1 km ≈ 1300 steps. NEVER return steps: 0 if distance given.
- If user says "total 5000 steps today" with no previous entry, just create one entry with 5000.

EXERCISE RULES:
- Each exercise entry is ADDITIVE.
- Include exercises array with name, sets, reps.
- Estimate calories burned: pushups 3x10 ≈ 15 cal, crunches 3x15 ≈ 20 cal, squats 3x15 ≈ 30 cal, running 30 min ≈ 300 cal, yoga 30 min ≈ 120 cal.
- Walking/running with distance → use "activity" type. Gym/bodyweight → use "exercise" type.

MULTI-ENTRY RULE:
- A single log can produce MULTIPLE entries. "ate biryani, walked 2 km, spent ₹350" → food + activity + expense.
- Parse ALL activities from the text. Do not ignore any.

GENERAL FALLBACK:
- If text doesn't match any category clearly, return type "general" with entries: [{"type":"general"}]
- NEVER return an empty entries array.`;

      const response = await callGemini(prompt);
      const match = response.match(/\{[\s\S]*\}/);
      if (match) {
        const parsed = JSON.parse(match[0]);
        return {
          entries: Array.isArray(parsed.entries) ? parsed.entries : [],
          summary: parsed.summary || '',
        };
      }
    } catch (e) { console.error('classify error:', e); }
    return mockClassify(text);
  },

  async analyzeFood(entries: FoodEntry[]) {
    if (!hasGeminiApiKey() || entries.length === 0) return mockFoodAnalysis(entries);
    try {
      const list = entries.map(e => `${e.name} (${e.portionSize})${e.calories ? ` ${e.calories}cal` : ''}`).join(', ');
      const prompt = `You are a nutritionist. Analyze this day's food intake:
${list}

Return ONLY JSON:
{
  "estimatedCalories": <total>,
  "protein": <total grams>,
  "carbs": <total grams>,
  "fat": <total grams>,
  "summary": "Brief nutritional assessment",
  "suggestions": ["2-3 specific suggestions for remaining meals today"]
}

Use standard Indian food nutrition data. Be specific about what to eat next.`;
      const response = await callGemini(prompt);
      const match = response.match(/\{[\s\S]*\}/);
      if (match) return JSON.parse(match[0]);
    } catch (e) { console.error('food analysis error:', e); }
    return mockFoodAnalysis(entries);
  },

  async search(query: string, context = ''): Promise<string> {
    if (!hasGeminiApiKey()) return mockSearch(query);
    try {
      const prompt = `You are LifeLog AI — a personal health, fitness, and life tracking assistant.
Answer based on the user's actual data. Be specific, actionable, encouraging.
IMPORTANT: Do NOT use markdown formatting. No ** for bold, no # headers, no * for bullets. Use plain text with emoji and line breaks. Use "•" for bullet points.
${context ? `\nUser's data:\n${context}\n` : ''}
Question: ${query}`;
      return await callGemini(prompt);
    } catch (e) {
      return `⚠️ AI unavailable: ${e instanceof Error ? e.message : 'error'}`;
    }
  },

  async extractText(imageBase64: string): Promise<string> {
    if (!hasGeminiApiKey()) return 'OCR requires a Gemini API key.';
    try {
      return await callGeminiVision('Extract all text from this image. If it is a receipt/bill, identify total amount, date, vendor name, and list all line items with prices.', imageBase64);
    } catch (e) {
      return `OCR failed: ${e instanceof Error ? e.message : 'error'}`;
    }
  },

  async generateInsights(data: {
    sleepHours: number; steps: number; expenses: number;
    meals: number; calories: number;
    exerciseLog?: string; foodLog?: string;
  }): Promise<string[]> {
    if (!hasGeminiApiKey()) return mockInsights(data);
    try {
      const prompt = `You are a professional health coach. Analyze this user's day and give 3-4 actionable insights.

Today's data:
- Sleep: ${data.sleepHours > 0 ? data.sleepHours + 'h' : 'not logged'}
- Steps: ${data.steps > 0 ? data.steps.toLocaleString() : 'not logged'}
- Meals: ${data.meals > 0 ? data.meals + ' meals, ' + data.calories + ' cal' : 'not logged'}
- Spending: ${data.expenses > 0 ? '₹' + data.expenses : 'none'}
${data.exerciseLog ? `- Exercise: ${data.exerciseLog}` : '- Exercise: not logged'}
${data.foodLog ? `- Food details: ${data.foodLog}` : ''}

RULES:
- If something is "not logged", say "Not tracked yet — log it for better insights" 
- Do NOT give negative/harsh advice for missing data
- For exercise: estimate total calories burned
- For food: assess if diet is balanced, suggest what to eat next
- For sleep: assess quality based on hours
- Each insight 1-2 sentences, with emoji prefix
- Be encouraging and specific

Return ONLY a JSON array of strings.`;
      const response = await callGemini(prompt);
      const match = response.match(/\[[\s\S]*\]/);
      if (match) return JSON.parse(match[0]);
    } catch (e) { console.error('insights error:', e); }
    return mockInsights(data);
  },
};

function mockClassify(text: string): ClassifyResult {
  const l = text.toLowerCase();
  const entries: ClassifiedEntry[] = [];

  // Sleep: try to extract hours from "X hours" or "slept for Xh"
  if (/slept|sleep|bed|woke|nap/.test(l)) {
    const hoursMatch = l.match(/(\d+\.?\d*)\s*(?:hours?|hrs?|h\b)/);
    entries.push({ type: 'sleep', sleepHours: hoursMatch ? parseFloat(hoursMatch[1]) : undefined });
  }
  // Expense: extract amount
  if (/spent|paid|₹|rs\s?\d|bought|cost|\d+\s*rupee/.test(l)) {
    const amtMatch = l.match(/(?:₹|rs\.?\s*)(\d+)/);
    entries.push({ type: 'expense', amount: amtMatch ? parseInt(amtMatch[1]) : undefined });
  }
  // Food
  if (/ate|food|meal|breakfast|lunch|dinner|snack|drank|coffee|tea|chana|rice|roti/.test(l)) {
    entries.push({ type: 'food' });
  }
  // Activity: extract km or steps
  if (/walk|run|jog|km|steps/.test(l)) {
    const kmMatch = l.match(/(\d+\.?\d*)\s*km/);
    const stepsMatch = l.match(/(\d+)\s*steps/);
    const km = kmMatch ? parseFloat(kmMatch[1]) : undefined;
    const steps = stepsMatch ? parseInt(stepsMatch[1]) : (km ? Math.round(km * 1300) : undefined);
    entries.push({ type: 'activity', steps, distanceKm: km });
  }
  // Exercise
  if (/exercise|pushup|crunch|yoga|workout|situp|plank|squat|gym|sets|reps/.test(l)) {
    entries.push({ type: 'exercise' });
  }
  if (entries.length === 0) entries.push({ type: 'general' });
  return { entries, summary: text.slice(0, 80) };
}

function mockFoodAnalysis(entries: FoodEntry[]) {
  const cal = entries.reduce((s, e) => s + (e.calories || 0), 0);
  return { estimatedCalories: cal, summary: `${entries.length} items, ~${cal} cal.`, suggestions: ['Add more vegetables.', 'Drink 8 glasses of water.', 'Include protein each meal.'] };
}

function mockSearch(q: string): string {
  if (/expense|spent|money/.test(q.toLowerCase())) return 'Track expenses in Today tab. See Dashboard for breakdown.';
  if (/sleep/.test(q.toLowerCase())) return 'Log sleep daily. Aim for 7-9 hours.';
  if (/food|diet|cal/.test(q.toLowerCase())) return 'Use Food quick-add for meals.';
  return 'Ask me about expenses, sleep, food, or fitness!';
}

function mockInsights(d: { sleepHours: number; steps: number; expenses: number; meals: number; calories: number }): string[] {
  const r: string[] = [];
  if (d.sleepHours > 0) r.push(d.sleepHours >= 7 ? `✅ ${d.sleepHours}h sleep — well rested!` : `💤 ${d.sleepHours}h — aim for 7-9h.`);
  else r.push('😴 Sleep not tracked yet — log it!');
  if (d.steps > 0) r.push(d.steps >= 10000 ? `🎉 ${d.steps.toLocaleString()} steps!` : `🚶 ${d.steps.toLocaleString()} steps. ${(10000 - d.steps).toLocaleString()} to 10k.`);
  if (d.meals > 0) r.push(`🍽️ ${d.meals} meals, ~${d.calories} cal.`);
  if (d.expenses > 0) r.push(`💰 ₹${d.expenses.toLocaleString('en-IN')} spent.`);
  if (r.length === 0) r.push('📝 Start logging to get insights!');
  return r;
}

File: ios-app/LifeLogAI/Services/GeminiService.swift
//
//  GeminiService.swift
//  LifeLogAI
//
//  Integrates with Google's Gemini API for:
//  - Classifying free-text log entries into structured data
//  - Analyzing food intake and providing nutritional insights
//  - OCR/text extraction from images and PDFs
//  - Answering search queries about user data or general facts
//
//  API Documentation: https://ai.google.dev/docs
//  Free Tier: 15 requests/minute, 1M tokens/month
//

import Foundation

class GeminiService {
    static let shared = GeminiService()
    
    private let apiKey = Config.geminiAPIKey
    private let model = Config.geminiModel
    private let baseURL = Config.geminiBaseURL
    
    private init() {}
    
    // MARK: - API Request
    
    private func makeRequest(prompt: String) async throws -> String {
        let endpoint = "\(baseURL)/models/\(model):generateContent?key=\(apiKey)"
        
        guard let url = URL(string: endpoint) else {
            throw GeminiError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "contents": [
                [
                    "parts": [
                        ["text": prompt]
                    ]
                ]
            ],
            "generationConfig": [
                "temperature": 0.7,
                "maxOutputTokens": 1024
            ]
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw GeminiError.invalidResponse
        }
        
        guard httpResponse.statusCode == 200 else {
            throw GeminiError.apiError(statusCode: httpResponse.statusCode)
        }
        
        // Parse response
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let candidates = json["candidates"] as? [[String: Any]],
              let firstCandidate = candidates.first,
              let content = firstCandidate["content"] as? [String: Any],
              let parts = content["parts"] as? [[String: Any]],
              let firstPart = parts.first,
              let text = firstPart["text"] as? String else {
            throw GeminiError.parsingError
        }
        
        return text
    }
    
    // MARK: - Log Classification
    
    /// Classify a free-text log entry into structured data
    /// Returns a dictionary with type, confidence, and extracted data
    func classifyLog(text: String) async -> ClassificationResult {
        // Check if API key is configured
        guard apiKey != "YOUR_GEMINI_API_KEY_HERE" else {
            // Return mock data for demo
            return mockClassifyLog(text: text)
        }
        
        let prompt = """
        Analyze this personal log entry and classify it. Return a JSON object with:
        - type: one of "general", "expense", "food", "sleep", "exercise", "note"
        - confidence: a number between 0 and 1
        - extractedData: any relevant structured data you can extract
        
        Log entry: "\(text)"
        
        Respond with only the JSON object, no markdown.
        """
        
        do {
            let response = try await makeRequest(prompt: prompt)
            // Parse the JSON response
            if let data = response.data(using: .utf8),
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                let type = json["type"] as? String ?? "general"
                let confidence = json["confidence"] as? Double ?? 0.5
                let extractedData = json["extractedData"] as? [String: Any] ?? [:]
                return ClassificationResult(type: type, confidence: confidence, extractedData: extractedData)
            }
        } catch {
            print("GeminiService classifyLog error: \(error)")
        }
        
        return mockClassifyLog(text: text)
    }
    
    /// Mock classification for demo mode
    private func mockClassifyLog(text: String) -> ClassificationResult {
        let lower = text.lowercased()
        
        if lower.contains("spent") || lower.contains("paid") || lower.contains("$") || lower.contains("bought") {
            return ClassificationResult(type: "expense", confidence: 0.85, extractedData: ["suggestedCategory": "General"])
        }
        if lower.contains("ate") || lower.contains("food") || lower.contains("meal") || lower.contains("breakfast") || lower.contains("lunch") || lower.contains("dinner") {
            return ClassificationResult(type: "food", confidence: 0.80, extractedData: ["suggestedMealType": "meal"])
        }
        if lower.contains("slept") || lower.contains("sleep") || lower.contains("bed") || lower.contains("woke") {
            return ClassificationResult(type: "sleep", confidence: 0.90, extractedData: [:])
        }
        if lower.contains("walk") || lower.contains("run") || lower.contains("exercise") || lower.contains("steps") || lower.contains("gym") {
            return ClassificationResult(type: "exercise", confidence: 0.88, extractedData: [:])
        }
        
        return ClassificationResult(type: "general", confidence: 0.50, extractedData: [:])
    }
    
    // MARK: - Food Analysis
    
    /// Analyze food entries and return nutritional insights
    func analyzeFood(entries: [FoodEntry]) async -> FoodAnalysis {
        guard apiKey != "YOUR_GEMINI_API_KEY_HERE" else {
            return mockFoodAnalysis(entries: entries)
        }
        
        let foodList = entries.map { "\($0.name) (\($0.portionSize))" }.joined(separator: ", ")
        
        let prompt = """
        Analyze these food items eaten today and provide nutritional insights:
        \(foodList)
        
        Return a JSON object with:
        - estimatedCalories: total estimated calories
        - summary: a brief summary of the diet
        - suggestions: array of 2-3 health suggestions
        
        Respond with only the JSON object, no markdown.
        """
        
        do {
            let response = try await makeRequest(prompt: prompt)
            if let data = response.data(using: .utf8),
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                return FoodAnalysis(
                    estimatedCalories: json["estimatedCalories"] as? Int ?? 0,
                    summary: json["summary"] as? String ?? "",
                    suggestions: json["suggestions"] as? [String] ?? []
                )
            }
        } catch {
            print("GeminiService analyzeFood error: \(error)")
        }
        
        return mockFoodAnalysis(entries: entries)
    }
    
    private func mockFoodAnalysis(entries: [FoodEntry]) -> FoodAnalysis {
        let totalCalories = entries.compactMap { $0.calories }.reduce(0, +)
        return FoodAnalysis(
            estimatedCalories: totalCalories,
            summary: "You logged \(entries.count) food items today with approximately \(totalCalories) calories.",
            suggestions: [
                "Try to include more vegetables in your meals.",
                "Stay hydrated — aim for 8 glasses of water.",
                "Consider adding protein to each meal for sustained energy."
            ]
        )
    }
    
    // MARK: - Search / Chat
    
    /// Answer a search query about user data or general facts
    func search(query: String, context: String = "") async -> String {
        guard apiKey != "YOUR_GEMINI_API_KEY_HERE" else {
            return mockSearch(query: query)
        }
        
        let prompt = """
        You are a helpful AI assistant for a personal life logging app called LifeLog AI.
        The user is asking a question. If it's about their personal data, use the context provided.
        If it's a general factual question, answer it helpfully.
        
        User's data context:
        \(context)
        
        User's question: \(query)
        
        Provide a helpful, concise response.
        """
        
        do {
            return try await makeRequest(prompt: prompt)
        } catch {
            print("GeminiService search error: \(error)")
            return mockSearch(query: query)
        }
    }
    
    private func mockSearch(query: String) -> String {
        let lower = query.lowercased()
        
        if lower.contains("expense") || lower.contains("spent") || lower.contains("money") {
            return "Based on your logs, you've been tracking expenses consistently. To get detailed spending insights, make sure to log each expense with a category. Once the Gemini API is connected with your key, I'll provide personalized spending analysis!"
        }
        if lower.contains("sleep") {
            return "Sleep is important! The WHO recommends 7-9 hours for adults. Log your sleep daily to track patterns. With the Gemini API connected, I can analyze your sleep quality trends and provide personalized tips."
        }
        if lower.contains("food") || lower.contains("diet") || lower.contains("eat") || lower.contains("calories") {
            return "Good nutrition is key to wellbeing. Log your meals with portion sizes and estimated calories for better tracking. Once integrated with Gemini, I can provide nutritional analysis and meal suggestions."
        }
        if lower.contains("exercise") || lower.contains("steps") || lower.contains("fitness") || lower.contains("walk") {
            return "Great job staying active! The WHO recommends at least 150 minutes of moderate exercise weekly. Log your steps and activities to track progress. Enable HealthKit for automatic step tracking!"
        }
        
        return """
        That's a great question! I'm your LifeLog AI assistant. 
        
        To unlock full AI capabilities:
        1. Get your free Gemini API key at makersuite.google.com
        2. Add it to Config.swift
        3. Rebuild the app
        
        Then I can answer questions about your personal data, provide insights, and help with general knowledge queries!
        
        For now, try asking about: expenses, sleep, food, or exercise.
        """
    }
    
    // MARK: - OCR / Text Extraction
    
    /// Extract text from an image using Gemini's vision capabilities
    func extractText(from imageData: Data) async -> String {
        guard apiKey != "YOUR_GEMINI_API_KEY_HERE" else {
            return "OCR extraction requires a Gemini API key. Get your free key at makersuite.google.com and add it to Config.swift. Once configured, I can read text from receipts, bills, documents, and more!"
        }
        
        // For vision, we need to use gemini-1.5-flash with image
        let endpoint = "\(baseURL)/models/gemini-1.5-flash:generateContent?key=\(apiKey)"
        
        guard let url = URL(string: endpoint) else {
            return "Error: Invalid API URL"
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let base64Image = imageData.base64EncodedString()
        
        let body: [String: Any] = [
            "contents": [
                [
                    "parts": [
                        ["text": "Extract all text from this image. If it's a receipt or bill, also identify the total amount. Format the response clearly."],
                        [
                            "inline_data": [
                                "mime_type": "image/jpeg",
                                "data": base64Image
                            ]
                        ]
                    ]
                ]
            ]
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            let (data, _) = try await URLSession.shared.data(for: request)
            
            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let candidates = json["candidates"] as? [[String: Any]],
               let firstCandidate = candidates.first,
               let content = firstCandidate["content"] as? [String: Any],
               let parts = content["parts"] as? [[String: Any]],
               let firstPart = parts.first,
               let text = firstPart["text"] as? String {
                return text
            }
        } catch {
            print("GeminiService OCR error: \(error)")
        }
        
        return "Could not extract text from image. Please try again."
    }
}

// MARK: - Supporting Types

struct ClassificationResult {
    let type: String
    let confidence: Double
    let extractedData: [String: Any]
}

struct FoodAnalysis {
    let estimatedCalories: Int
    let summary: String
    let suggestions: [String]
}

enum GeminiError: LocalizedError {
    case invalidURL
    case invalidResponse
    case apiError(statusCode: Int)
    case parsingError
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .invalidResponse:
            return "Invalid response from server"
        case .apiError(let code):
            return "API error: \(code)"
        case .parsingError:
            return "Could not parse response"
        }
    }
}
